### DISCUSSION TITLE
Equivalent of `subscription-transport-ws` `SubscriptionClient.unsubscribeAll()`?

### DISCUSSION THREAD

**Discussion by KeithGillette** [enisdenjo/graphql-ws#363] [Q&A] [answered]
In moving from the no-longer-maintained `subscription-transport-ws` library with Apollo Client, I'm struggling with how to gracefully close all Subscriptions on user sign-out. Previously, we would call  `subscription-transport-ws`'s `SubscriptionClient.unsubscribeAll()` method on sign-out. Is there an equivalent method in `graphql-ws`? If not, how might one ensure that all GraphQL Subscriptions are completed and the WebSocket connection closed on user sign-out?

**Comment by enisdenjo** [comment, marked as answer]
You should use the [`Client.dispose()` method](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/client.Client.md#dispose). It will complete all active subscriptions, disconnect the WebSocket and clean up resources.

**Reply by KeithGillette** [reply]
Thanks for your reply, @enisdenjo. I tried `Client.dispose()` but it seemed to do too much, as when the user signs back in after a call to `Client.dispose()`, I do not see new Subscriptions starting. The `subscription-transport-ws` `SubscriptionClient.unsubscribeAll()` method just completed active subscriptions and closed the connection but left the WebSocket client ready to start a new session. Is there something that needs to be done to re-initialize `graphql-ws` after `Client.dispose()`?

**Reply by enisdenjo** [reply]
The dispose method disposes the client and it is therefore completely unusable afterwards. You have to create a new client if you want to connect again, a client factory could be one approach - roughly:
```ts
import { Client, createClient } from 'graphql-ws';

let activeClient: Client | null = null;
function getClient(): Client {
  if (!activeClient) {
    const client = createClient({ ... });
    activeClient = {
      ...client,
      dispose: () => {
        activeClient = null;
        client.dispose();
      };
    };
  }
  return activeClient;
}
```

**Reply by KeithGillette** [reply]
Thanks so much for the client factory code sample, @enisdenjo. I have adapted it to our project and it appears to work perfectly.
